File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
  if (!words.includes(target)) {
    return 0;
  }
  let answer = Infinity;
  const visited = Array(words.length).fill(false);
  function dfs(start, count) {
    if (start === target) {
      answer = Math.min(answer, count);
      return;
    }
    for (let i = 0; i < words.length; i++) {
      if (visited[i]) continue;
      if (checkChange(words[i], start)) {
        // console.log('hi?', start, words[i]);
        visited[i] = true;
        dfs(words[i], count + 1);
        visited[i] = false;
      }
    }
  }
  dfs(begin, 0);
  return answer;
}

function checkChange(word1, word2) {
  let count = 0;
  for (let i = 0; i < word1.length; i++) {
    if (word1[i] !== word2[i]) {
      count++;
    }
    if (count > 1) {
      return false;
    }
  }
  return true;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    if target not in words:
        return answer
    visited = {word: False for word in words}

    def bfs():
        q = deque([(begin, 0)])

        while q:
            current_word, change_cnt = q.popleft()

            if current_word == target:
                return change_cnt

            for word in words:
                if not visited[word] and findCanChange(word, current_word):
                    visited[word] = True
                    q.append((word, change_cnt + 1))

        return 0

    return bfs()

def findCanChange(word1, word2):
    diff = 0

    for i in range(len(word1)):
        if word1[i] != word2[i]:
            diff += 1

    return diff == 1